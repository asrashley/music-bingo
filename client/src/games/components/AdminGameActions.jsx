import React from 'react';
import PropTypes from 'prop-types';
import { saveAs } from 'file-saver';

import {
  BusyDialog, ConfirmDialog, FileDialog, ProgressDialog
} from '../../components';

import { addMessage } from '../../messages/messagesSlice';
import { importDatabase } from '../../admin/adminSlice';
import {
  deleteGame, fetchGamesIfNeeded, importGame, invalidateGames
} from '../gamesSlice';

import { api } from '../../endpoints';

import { GamePropType } from '../types/Game';

export function AdminActionPanel({ deleteGame, exportGame, importGame, game }) {
  return (
    <div className="action-panel">
      <button className="btn btn-primary ml-2" onClick={importGame}>
        Import a game
      </button>
      {(game && exportGame) && <button className="btn btn-primary ml-2"
        onClick={exportGame}>Export game</button>}
      {(game && deleteGame) && <button className="btn btn-danger ml-2"
        onClick={deleteGame}>Delete game</button>}
    </div>
  );
}

AdminActionPanel.propTypes = {
  game: GamePropType,
  deleteGame: PropTypes.func,
  exportGame: PropTypes.func,
  importGame: PropTypes.func.isRequired,
};

function ErrorMessage({ error }) {
  if (!error) {
    return null;
  }
  return (
    <div className="alert alert-warning" role="alert">
      <span className="error-message">{error}</span>
    </div>
    );
}
ErrorMessage.propTypes = {
  error: PropTypes.string
};

export class AdminGameActions extends React.Component {
  static DATABASE_IMPORT = 1;
  static GAME_IMPORT = 2;

  static propTypes = {
    databaseImporting: PropTypes.object,
    dispatch: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired,
    game: GamePropType,
    gameImporting: PropTypes.object,
    importing: PropTypes.object,
  };

  state = {
    ActiveDialog: null,
    dialogData: null,
    importType: 0,
    error: null
  };

  render() {
    const { ActiveDialog, dialogData, error } = this.state;
    const { children, game, importing } = this.props;

    return (
      <React.Fragment>
        <ErrorMessage error={error} />
        {children}
        <AdminActionPanel
          game={game}
          deleteGame={this.confirmDelete}
          exportGame={this.exportGame}
          importGame={this.onClickImportGame}
        />;
        {ActiveDialog && <ActiveDialog backdrop {...dialogData} {...importing} /> }
      </React.Fragment>);
  }

  confirmDelete = (ev) => {
    ev.preventDefault();
    const { game } = this.props;
    this.setState({
      ActiveDialog: ConfirmDialog,
      dialogData: {
        changes: [
          `Delete group ${game.id}`
        ],
        title: "Confirm delete game",
        onCancel: this.cancelDialog,
        onConfirm: this.deleteGame,
      }
    });
    return false;
  };

  cancelDialog = () => {
    this.setState({
      ActiveDialog: null,
      dialogData: null,
      importType: 0,
    });
  };

  deleteGame = () => {
    const { game, dispatch, onDelete } = this.props;
    this.setState({ ActiveDialog: null, dialogData: null });
    dispatch(deleteGame(game)).then(result => {
      const { payload } = result;
      if (payload?.error || result.error) {
        this.setState({ error: (payload.error ? payload.error : result.error) });
      } else {
        onDelete(game);
      }
    });
  };

  onClickImportDatabase = () => {
    this.setState({
      ActiveDialog: FileDialog,
      dialogData: {
        title: 'Select a json database file to import',
        accept: '.json,application/json',
        submit: 'Import database',
        onCancel: this.cancelDialog,
        onFileUpload: this.onFileSelected
      },
      importType: AdminGameActions.DATABASE_IMPORT
    });
  };

  onChangeReplaceDate = (e) => {
    const { dialogData } = this.state;
    this.setState({
      dialogData: {
        ...dialogData,
        replaceDate: e.target.checked
      }
    });
  }

  onClickImportGame = () => {
    const extraFields = <div className="form-control">
      <label htmlFor="replaceDate" className="replace-date-label">Replace game start date during import?</label>
      <input type="checkbox" name="replaceDate" defaultChecked onChange={this.onChangeReplaceDate}></input>
    </div>;

    this.setState({
      ActiveDialog: FileDialog,
      dialogData: {
        title: 'Select a gameTracks.json file to import',
        accept: '.json,application/json',
        submit: 'Import game',
        onCancel: this.cancelDialog,
        onFileUpload: this.onFileSelected,
        replaceDate: true,
        extraFields
      },
      importType: AdminGameActions.GAME_IMPORT
    });
  };

  importComplete = () => {
    const { databaseImporting, gameImporting, dispatch } = this.props;
    const { importType } = this.state;
    if (importType === AdminGameActions.DATABASE_IMPORT && databaseImporting?.done === true) {
        document.location.reload();
        return;
    }
    if (importType === AdminGameActions.GAME_IMPORT && gameImporting?.done === true) {
      dispatch(invalidateGames());
      dispatch(fetchGamesIfNeeded());
    }
    this.setState({
      ActiveDialog: null,
      dialogData: null,
      importType: 0,
    });
  };

  exportDatabase = () => {
    const { dispatch } = this.props;
    this.setState(() => ({
      ActiveDialog: BusyDialog,
      dialogData: {
        onClose: this.cancelDialog,
        title: "Exporting database",
        text: "Please wait, exporting database...",
      }
    }), () => {
      dispatch(api.exportDatabase()).then(response => {
        const filename = `database-${Date.now()}.json`;
        return response.payload.blob().then(blob => {
          this.setState({ ActiveDialog: null, dialogData: null });
          saveAs(blob, filename);
          return filename;
        });
      });
    });
  };

  exportGame = () => {
    const { dispatch, game } = this.props;
    dispatch(api.exportGame({
      gamePk: game.pk
    })).then(response => {
      const filename = `game-${game.id}.json`;
      return response.payload.blob().then(blob => {
        saveAs(blob, filename);
        return filename;
      });
    });
  };

  onFileSelected = (file) => {
    var reader = new FileReader();
    reader.onload = (ev) => this.onFileLoaded(file.name, ev);
    reader.onerror = (err) => {
      this.setState({ activeDialog: null, activeDialogData: null });
      addMessage({ type: "error", text: err });
    };
    reader.readAsText(file);
  };

  onFileLoaded(filename, event) {
    const { dispatch } = this.props;
    const { importType } = this.state;
    const { replaceDate } = this.state.dialogData;
    const title = importType === AdminGameActions.GAME_IMPORT ?
      `Importing game from "${filename}"` :
      `Importing database from "${filename}"`;

    this.setState({
      ActiveDialog: ProgressDialog,
      dialogData: {
        title,
        onCancel: this.cancelDialog,
        onClose: this.importComplete
      }
    });
    const data = JSON.parse(event.target.result);
    if (importType === AdminGameActions.GAME_IMPORT) {
      if (replaceDate === true && 'Games' in data) {
        const start = new Date().toISOString();
        const end = new Date(Date.now() + 12*3600000).toISOString();
        data.Games = data.Games.map(game => ({
          ...game,
          start,
          end
        }));
      }
      dispatch(importGame(filename, data));
    } else {
      dispatch(importDatabase(filename, data));
    }
  }
}


