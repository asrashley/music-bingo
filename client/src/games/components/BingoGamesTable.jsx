import React from 'react';
import { reverse } from 'named-urls';
import routes from '../../routes';
import { Link } from 'react-router-dom';

function formatTime(value) {
  let hours = value.getHours();
  const ampm = value.getHours() >= 12 ? 'pm' : 'am';
  hours = hours % 12;
  if (hours === 0) {
    hours = 12;
  }
  //hours = `0${hours}`.slice(-2);
  const minutes = `0${value.getMinutes()}`.slice(-2);
  return `${hours}:${minutes} ${ampm}`;
}

const TableRow = ({ game, past }) => {
  let ticketUrl, themeUrl, linkText;

  if (past) {
    linkText = 'View track listing';
    ticketUrl = themeUrl = reverse(`${routes.trackListing}`, { gameId: game.id });
  } else if (game.userCount === 0) {
    linkText = "Choose tickets";
    ticketUrl = reverse(`${routes.game}`, { gameId: game.id });
    themeUrl = reverse(`${routes.game}`, { gameId: game.id });
  } else {
    ticketUrl = reverse(`${routes.play}`, { gameId: game.id });
    themeUrl = reverse(`${routes.game}`, { gameId: game.id });
    linkText = `You have chosen ${game.userCount} ticket${(game.userCount > 1) ? 's' : ''}`;
  }
  const start = new Date(game.start);
  return (
    <React.Fragment>
      {game.round === 1 && (<tr>
        <th colSpan="4">{start.toDateString()}</th>
      </tr>)}
      <tr className={`round-${game.round}`}>
        <td className="round-column">{game.round}</td>
        <td className="date-column">{formatTime(start)}</td>
        <td className="theme-column"><Link to={themeUrl}>{game.title}</Link></td>
        <td className="ticket-column"><Link to={ticketUrl}>{linkText}</Link></td>
      </tr>
    </React.Fragment>
  );
};

export const BingoGamesTable = ({ title, games, past, onReload }) => {
  return (
    <table className="table table-striped table-bordered game-list">
      <thead>
        <tr>
          <th colSpan="4" className="available-games">{title}
            <button className="btn btn-light refresh-icon btn-sm" onClick={onReload}>&#x21bb;</button>
          </th>
        </tr>
        <tr>
          <th className="round-column">Round</th>
          <th className="date-column">Date and Time</th>
          <th className="theme-column">Theme</th>
          {past && <th className="ticket-column">Track listing</th>}
          {!past && <th className="ticket-column">Your Tickets</th>}
        </tr>
      </thead>
      <tbody>
        {games.map((game, idx) => (<TableRow game={game} key={idx} past={past} />))}
      </tbody>
    </table>
  );
};
