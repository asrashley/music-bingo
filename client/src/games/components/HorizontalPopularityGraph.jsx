import React from 'react';
import PropTypes from 'prop-types';

import { ThemeRow } from './ThemeRow';
import { ThemePropType } from '../types/Theme';

export function HorizontalPopularityGraph({ popularity, onRotate }) {
  return (
    <table className="horiz-popularity-graph">
      <thead>
        <tr>
          <th className="title">Theme</th>
          <th className="count">Popularity
            <button className="btn btn-light rotate-icon" onClick={onRotate}>&nbsp;</button>
          </th>
        </tr>
      </thead>
      <tbody>
        {popularity.map((theme, idx) => <ThemeRow key={idx} index={idx} theme={theme} />)}
      </tbody>
    </table>
  );
}

HorizontalPopularityGraph.propTypes = {
  popularity: PropTypes.arrayOf(ThemePropType).isRequired,
  onRotate: PropTypes.func.isRequired
};
