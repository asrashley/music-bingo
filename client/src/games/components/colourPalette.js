
export const colours = [
  '#FF0000',
  '#008000',
  '#00EFEF',
  '#808000',
  '#BFBF00',
  '#808080',
  '#008080',
  '#00EF00',
  '#C0C0C0',
  '#800000',
  '#0000FF',
  '#000080',
  '#FF00FF',
  '#800080',
];
