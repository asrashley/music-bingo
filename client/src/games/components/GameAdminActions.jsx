import React from 'react';
import { useDispatch } from 'react-redux';
import { AdminActions } from '../../admin/components/AdminActions';
import { fetchGamesIfNeeded, invalidateGames } from '../gamesSlice';

export function GameAdminActions() {
    const dispatch = useDispatch();
    const onReload = () => {
        dispatch(invalidateGames());
        dispatch(fetchGamesIfNeeded());
    };

    return (
        <AdminActions alwaysShowChildren={true}>
            <button
                className="btn btn-primary"
                onClick={onReload}>
                Reload
            </button>
        </AdminActions>
    );
}
