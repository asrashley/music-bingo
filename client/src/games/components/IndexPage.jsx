import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { reverse } from 'named-urls';

import { initialState } from '../../app/initialState';
import { fetchUserIfNeeded, userIsLoggedIn } from '../../user/userSlice';
import { fetchGamesIfNeeded } from '../gamesSlice';
import routes from '../../routes';
import { LoginDialog } from '../../user/components/LoginDialog';

import '../styles/games.scss';

const utcFields = ['getUTCHours', 'getUTCMinutes',
  'getUTCSeconds', 'getUTCDate',
  'getUTCMonth', 'getUTCFullYear'];

const localFields = utcFields.map(f => f.replace('UTC', ''));


const DateTime = ({ date, useUTC=false, withTimezone=false, withDate=true, withTime=true, ...params }) => {
  if (typeof (date) === "string") {
    date = new Date(date);
  };
  let fields = (useUTC === false) ? localFields : utcFields;
  fields = fields.map((f) => {
    const val = date[f]();
    if (/Year/.test(f)) {
      return `${val}`;
    }
    if (/Month/.test(f)) {
      return `0${val + 1}`.slice(-2);
    }
    return `0${val}`.slice(-2);
  });
  let tz = '';
  if (withTimezone === true) {
    if (useUTC === false) {
      const offset = date.getTimezoneOffset();
      if (offset === 0) {
        tz = ' GMT';
      } else {
        tz = ' BST';
      }
    } else {
      tz = ' UTC';
    }
  }
  const time = withTime ? `${fields[0]}:${fields[1]}:${fields[2]}` : '';
  const dstr = withDate ? `${fields[3]}/${fields[4]}/${fields[5]}` : '';
  if (!withTime) {
    tz = '';
  }
  return `${time} ${dstr} ${tz}`;
};

const TableRow = ({ game }) => {
  let url, linkText;

  if (game.userCount === 0) {
    linkText = "Choose tickets";
    url = reverse(`${routes.game}`, { gamePk: game.pk });
  } else {
    url = reverse(`${routes.play}`, { gamePk: game.pk });
    linkText = `You have chosen ${game.userCount} ticket${(game.userCount > 1) ? 's' : ''}`;
  }
  return (
    <tr>
      <td className="round-column">{game.gameRound}</td>
      <td className="date-column">
        <DateTime date={game.start} withTime={false} />
      </td>
      <td className="theme-column">
        <Link to={reverse(`${routes.game}`, { gamePk: game.pk })}>{game.title}</Link>
      </td>
      <td className="ticket-column">
        <Link to={url}>{linkText}</Link>
      </td>
    </tr>
  );
};

class IndexPage extends React.Component {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    loggedIn: PropTypes.bool.isRequired,
    user: PropTypes.object,
    games: PropTypes.array,
  };

  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(fetchUserIfNeeded());
    dispatch(fetchGamesIfNeeded());
  }

  componentWillReceiveProps(nextProps) {
    const { user, dispatch } = nextProps;
    if (user.pk !== this.props.user.pk) {
      dispatch(fetchGamesIfNeeded());
    }
  }

  render() {
    const { games, user, loggedIn } = this.props;
    return (
      <div id="games-page">
        {user && <h2 className="greeting">Hello {user.username}</h2>}
        <table className="table table-striped table-bordered game-list">
          <thead>
            <tr>
              <th colSpan="4" className="available-games">Available Bingo games</th>
            </tr>
            <tr>
              <th className="round-column">Round</th>
              <th className="date-column">Date</th>
              <th className="theme-column">Theme</th>
              <th className="ticket-column">Your Tickets</th>
            </tr>
          </thead>
          <tbody>
            {games.map((game, idx) => (<TableRow game={game} key={idx} />))}
          </tbody>
        </table>
        {!loggedIn && <LoginDialog dispatch={this.props.dispatch} onSuccess={() => null} />}
      </div>
    );
  }
}

//IndexPage.whyDidYouRender = true;

const mapStateToProps = (state, ownProps) => {
  state = state || initialState;
  const { user } = state;
  const { games, order } = state.games;
  const { location } = ownProps;
  const gameList = [];
  order.forEach(pk => gameList.push(games[pk]));
  return {
    loggedIn: userIsLoggedIn(state),
    user,
    location,
    games: gameList,
  };
};

IndexPage = connect(mapStateToProps)(IndexPage);

export {
  IndexPage
};
