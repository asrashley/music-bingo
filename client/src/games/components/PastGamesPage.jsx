import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { BingoGamesTable } from './BingoGamesTable';
import { AdminActions } from '../../admin/components/AdminActions';
import { PastGamesButtons } from './PastGamesButtons';

import { fetchUserIfNeeded } from '../../user/userSlice';
import { fetchGamesIfNeeded, invalidateGames } from '../gamesSlice';

import { getPastGamesList, getThemeSlug } from '../gamesSelectors';
import { getUser } from '../../user/userSelectors';

import { UserPropType } from '../../user/types/User';
import { GamePropType } from '../../games/types/Game';

import '../styles/games.scss';

class PastGamesPage extends React.Component {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    user: UserPropType.isRequired,
    pastGames: PropTypes.arrayOf(GamePropType).isRequired,
    slug: PropTypes.string,
  };

  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(fetchUserIfNeeded());
    dispatch(fetchGamesIfNeeded());
  }

  componentDidUpdate(prevProps) {
    const { user, dispatch } = this.props;
    if (user.pk > 0 && user.pk !== prevProps.user.pk) {
      dispatch(fetchGamesIfNeeded());
    }
  }

  onReload = () => {
    const { dispatch } = this.props;
    dispatch(invalidateGames());
    dispatch(fetchGamesIfNeeded());
  };

  render() {
    const { pastGames, user, slug } = this.props;
    let title = 'Previous Bingo games';
    if (slug) {
      if (pastGames.length) {
        title = `Previous "${pastGames[0].title}" Bingo games`;
      } else {
        title = `Previous "${slug}" Bingo games`;
      }
    }

    return (
      <div id="games-page" className={user.loggedIn ? '' : 'modal-open'}>
        <AdminActions alwaysShowChildren={true}>
          <button className="btn btn-primary"
            onClick={this.onReload}>Reload</button>
        </AdminActions>
        <PastGamesButtons page="all" />
        {user.groups?.guests === true && <div className="alert alert-info" role="alert">
          If you would like to see the track listing of every game, log out from this
          guest account and register an account.</div>}
        <BingoGamesTable games={pastGames} onReload={this.onReload} user={user} past
          title={title} slug={slug} />
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    user: getUser(state, ownProps),
    pastGames: getPastGamesList(state, ownProps),
    slug: getThemeSlug(state, ownProps),
  };
};

PastGamesPage = connect(mapStateToProps)(PastGamesPage);

export {
  PastGamesPage
};
