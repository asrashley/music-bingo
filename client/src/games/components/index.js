import { ListGamesPage } from './ListGamesPage';
import { TrackListing } from './TrackListing';
import { TrackListingPage } from './TrackListingPage';
import { ModifyGame } from './ModifyGame';
import { PastGamesPage } from './PastGamesPage';

export {
  ListGamesPage,
  ModifyGame,
  PastGamesPage,
  TrackListing,
  TrackListingPage,
};
