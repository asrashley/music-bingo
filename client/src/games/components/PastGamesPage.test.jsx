import React from 'react';
import { getByText, screen } from '@testing-library/react';
import log from 'loglevel';
import fetchMock from "fetch-mock-jest";

import { renderWithProviders, installFetchMocks } from '../../testHelpers';
import { formatDuration } from '../../components/DateTime';
import { PastGamesPage } from './PastGamesPage';

describe('PastGamesPage component', () => {
  beforeEach(() => {
    installFetchMocks(fetchMock, { loggedIn: true });
  });
  afterEach(() => {
    fetchMock.mockReset();
    log.resetLevel();
  });

  it('to render listing of previous games', async () => {
    //log.setLevel('debug');
    const { asFragment, store } = renderWithProviders(<PastGamesPage />);
    await screen.findAllByText("Rock & Power Ballads", { exact: false });
    const game = store.getState().games.games[159];
    game.tracks.forEach(track => {
      const tid = `track[${track.pk}]`;
      const dur = formatDuration(track.duration);
      expect('dur').not.toMatch(/N?aN?/);
      const row = screen.getByTestId(tid);
      expect(row).toBeVisible();
      expect(row).toHaveClass(`${game.options.colour_scheme}-theme`);
      getByText(row, track.title);
      getByText(row, track.artist);
      getByText(row, track.album);
    });
    expect(asFragment()).toMatchSnapshot();
  });
});