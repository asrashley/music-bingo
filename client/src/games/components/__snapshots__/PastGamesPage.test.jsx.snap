// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PastGamesPage component to render listing of previous games 1`] = `
<DocumentFragment>
  <div
    class=""
    id="games-page"
  >
    <div
      class="action-panel"
    >
      <button
        class="btn btn-primary ml-2"
      >
        Import a game
      </button>
    </div>
    <div
      class="vert-popularity-graph"
    >
      <button
        class="btn btn-light rotate-icon"
      >
         
      </button>
      <div
        class="theme"
        style="font-size: 1em;"
      >
        <div
          class="bar-wrap"
        >
          <div
            class="bar"
            style="height: 50%; background-color: rgb(255, 0, 0);"
          >
            1
          </div>
        </div>
        <div
          class="name"
          style="color: rgb(255, 0, 0);"
        >
          Pot Luck
        </div>
      </div>
      <div
        class="theme"
        style="font-size: 1em;"
      >
        <div
          class="bar-wrap"
        >
          <div
            class="bar"
            style="height: 100%; background-color: rgb(0, 128, 0);"
          >
            2
          </div>
        </div>
        <div
          class="name"
          style="color: rgb(0, 128, 0);"
        >
          Rock
        </div>
      </div>
      <div
        class="theme"
        style="font-size: 1em;"
      >
        <div
          class="bar-wrap"
        >
          <div
            class="bar"
            style="height: 50%; background-color: rgb(0, 239, 239);"
          >
            1
          </div>
        </div>
        <div
          class="name"
          style="color: rgb(0, 239, 239);"
        >
          Various Artists
        </div>
      </div>
      <div
        class="theme"
        style="font-size: 1em;"
      >
        <div
          class="bar-wrap"
        >
          <div
            class="bar"
            style="height: 50%; background-color: rgb(128, 128, 0);"
          >
            1
          </div>
        </div>
        <div
          class="name"
          style="color: rgb(128, 128, 0);"
        >
          Rock & Power Ballads
        </div>
      </div>
      <div
        class="theme"
        style="font-size: 1em;"
      >
        <div
          class="bar-wrap"
        >
          <div
            class="bar"
            style="height: 50%; background-color: rgb(191, 191, 0);"
          >
            1
          </div>
        </div>
        <div
          class="name"
          style="color: rgb(191, 191, 0);"
        >
          Motown
        </div>
      </div>
      <div
        class="theme"
        style="font-size: 1em;"
      >
        <div
          class="bar-wrap"
        >
          <div
            class="bar"
            style="height: 50%; background-color: rgb(128, 128, 128);"
          >
            1
          </div>
        </div>
        <div
          class="name"
          style="color: rgb(128, 128, 128);"
        >
          sunshine
        </div>
      </div>
      <div
        class="theme"
        style="font-size: 1em;"
      >
        <div
          class="bar-wrap"
        >
          <div
            class="bar"
            style="height: 50%; background-color: rgb(0, 128, 128);"
          >
            1
          </div>
        </div>
        <div
          class="name"
          style="color: rgb(0, 128, 128);"
        >
          Pride
        </div>
      </div>
    </div>
    <table
      class="table table-bordered game-list"
    >
      <thead>
        <tr>
          <th
            class="available-games"
            colspan="4"
          >
            Previous Bingo games
            <button
              aria-label="Reload"
              class="btn btn-light refresh-icon btn-sm"
            >
              ↻
            </button>
          </th>
        </tr>
        <tr>
          <th
            class="round-column"
          >
            Round
          </th>
          <th
            class="date-column"
          >
            Date and Time
          </th>
          <th
            class="theme-column"
          >
            Theme
          </th>
          <th
            class="ticket-column"
          >
            Track listing
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th
            colspan="4"
          >
            Thu Apr 05 2018
          </th>
        </tr>
        <tr
          class="round-1 green-theme"
          data-testid="pastgame[2]"
        >
          <td
            class="round-column"
          >
            1
          </td>
          <td
            class="date-column"
          >
            8:43:00 pm 05/04/2018
          </td>
          <td
            class="theme-column"
          >
            <a
              href="/history/18-04-25-2"
            >
              Pot Luck
            </a>
          </td>
          <td
            class="ticket-column"
          >
            <a
              href="/history/18-04-25-2"
            >
              View track listing
            </a>
          </td>
        </tr>
        <tr>
          <th
            colspan="4"
          >
            Fri Apr 06 2018
          </th>
        </tr>
        <tr
          class="round-1 blue-theme"
          data-testid="pastgame[1]"
        >
          <td
            class="round-column"
          >
            1
          </td>
          <td
            class="date-column"
          >
            7:24:25 pm 06/04/2018
          </td>
          <td
            class="theme-column"
          >
            <a
              href="/history/18-04-25-1"
            >
              Rock
            </a>
          </td>
          <td
            class="ticket-column"
          >
            <a
              href="/history/18-04-25-1"
            >
              View track listing
            </a>
          </td>
        </tr>
        <tr>
          <th
            colspan="4"
          >
            Sun Apr 22 2018
          </th>
        </tr>
        <tr
          class="round-1 blue-theme"
          data-testid="pastgame[159]"
        >
          <td
            class="round-column"
          >
            1
          </td>
          <td
            class="date-column"
          >
            3:43:56 pm 22/04/2018
          </td>
          <td
            class="theme-column"
          >
            <a
              href="/history/18-04-22-2"
            >
              Various Artists
            </a>
          </td>
          <td
            class="ticket-column"
          >
            <a
              href="/history/18-04-22-2"
            >
              View track listing
            </a>
          </td>
        </tr>
        <tr>
          <th
            colspan="4"
          >
            Thu May 03 2018
          </th>
        </tr>
        <tr
          class="round-1 blue-theme"
          data-testid="pastgame[3]"
        >
          <td
            class="round-column"
          >
            1
          </td>
          <td
            class="date-column"
          >
            7:51:00 pm 03/05/2018
          </td>
          <td
            class="theme-column"
          >
            <a
              href="/history/18-05-02-1"
            >
              Rock & Power Ballads
            </a>
          </td>
          <td
            class="ticket-column"
          >
            <a
              href="/history/18-05-02-1"
            >
              View track listing
            </a>
          </td>
        </tr>
        <tr
          class="round-2 green-theme"
          data-testid="pastgame[4]"
        >
          <td
            class="round-column"
          >
            2
          </td>
          <td
            class="date-column"
          >
            8:58:47 pm 03/05/2018
          </td>
          <td
            class="theme-column"
          >
            <a
              href="/history/18-05-02-2"
            >
              Motown
            </a>
          </td>
          <td
            class="ticket-column"
          >
            <a
              href="/history/18-05-02-2"
            >
              View track listing
            </a>
          </td>
        </tr>
        <tr>
          <th
            colspan="4"
          >
            Mon Jun 01 2020
          </th>
        </tr>
        <tr
          class="round-1 yellow-theme"
          data-testid="pastgame[84]"
        >
          <td
            class="round-column"
          >
            1
          </td>
          <td
            class="date-column"
          >
            7:00:00 pm 01/06/2020
          </td>
          <td
            class="theme-column"
          >
            <a
              href="/history/20-06-01-2"
            >
              Rock
            </a>
          </td>
          <td
            class="ticket-column"
          >
            <a
              href="/history/20-06-01-2"
            >
              View track listing
            </a>
          </td>
        </tr>
        <tr>
          <th
            colspan="4"
          >
            Thu Jun 04 2020
          </th>
        </tr>
        <tr
          class="round-1 yellow-theme"
          data-testid="pastgame[85]"
        >
          <td
            class="round-column"
          >
            1
          </td>
          <td
            class="date-column"
          >
            7:30:00 pm 04/06/2020
          </td>
          <td
            class="theme-column"
          >
            <a
              href="/history/20-06-03-1"
            >
              sunshine
            </a>
          </td>
          <td
            class="ticket-column"
          >
            <a
              href="/history/20-06-03-1"
            >
              View track listing
            </a>
          </td>
        </tr>
        <tr
          class="round-2 pride-theme"
          data-testid="pastgame[86]"
        >
          <td
            class="round-column"
          >
            2
          </td>
          <td
            class="date-column"
          >
            8:01:00 pm 04/06/2020
          </td>
          <td
            class="theme-column"
          >
            <a
              href="/history/20-06-03-2"
            >
              Pride
            </a>
          </td>
          <td
            class="ticket-column"
          >
            <a
              href="/history/20-06-03-2"
            >
              View track listing
            </a>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
